### PR TITLE
Add types declaration file, install typescript as devDependency

### PR DESCRIPTION
**Pull request by geopic**
The pre-commit linting process needed typescript in order to work with the types declaration file. Let me know if any issues arise!

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    SingleQueryBuilderType: AuthQueryBuilder<M, M>
+    NumberQueryBuilderType: AuthQueryBuilder<M, number>
+    PageQueryBuilderType: AuthQueryBuilder<M, Page<M>>
+  }
+  class AuthorizeModelClass extends Model {
+    QueryBuilderType: AuthQueryBuilder<this>
+    QueryBuilder: AuthQueryBuilder<this>
+  }
+
+  export default function authorize (
```
Comment (JaneJeon) [inline index.d.ts]: So... unfortunately this suffers from the same issue pointed out in https://github.com/JaneJeon/objection-hashid/pull/51#discussion_r373315460, where:
1. the type information is not actually usable from the outside world, and
2. custom query methods (specifically, `.authorize()` and `.action()`) are erased.

:/

**Comment by JaneJeon** [comment]
Oh, and as you can see in https://github.com/JaneJeon/objection-authorize/pull/79/checks?check_run_id=413223598 linting is still failing. @geopic

**Comment by geopic** [comment]
Sorry to hear it still isn't working right, I'll give it another look @JaneJeon .

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { Model, ModelClass, Page } from 'objection-2'
+import { Ability } from '@casl/ability'
+import { AccessControl } from 'role-acl-3'
```
[the remark this answers is not in the document]
Comment (geopic) [reply index.d.ts]: Noted :+1:

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  unauthenticatedErrorCode?: number
+  unauthorizedErrorCode?: number
+  userFromResult?: boolean
+  contextKey?: string
+  roleFromUser?: (user: Model) => string
+  resourceAugments?: { true: boolean; false: boolean; undefined: undefined }
+}
+
+declare module 'objection-authorize' {
+  class AuthQueryBuilder<M extends Model, R = M[]> extends Model.QueryBuilder<
```
[the remark this answers is not in the document]
Comment (geopic) [reply index.d.ts]: Will give this another shot soon based on the feedback from @olavim for objection-hashid.

**Comment by geopic** [comment]
I've pushed some new changes. Similar to `objection-hashid`, TS devs have to use `import` and then something like:

    const auth = authorize(acl, library, opts);
    class Post extends auth(Model) {
      ...
    }

If there are any details I've missed then let me know.

**Comment by JaneJeon** [comment]
Closed due to https://github.com/JaneJeon/objection-hashid/issues/46 (TLDR: it’s literally impossible :<)

**Pull request closed without merging**